server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java

```diff
@@ -1167,12 +1168,19 @@ public void writeNamedWriteableCollection(Collection<? extends NamedWriteable> l
         writeCollection(list, StreamOutput::writeNamedWriteable);
     }
 
+    private static <E extends Enum<E>> boolean assertEnumToWrite(E enumValue, TransportVersion version) {
+        assert enumValue instanceof XContentType == false : "XContentHelper#writeTo should be used for XContentType serialisation";
+        assert enumValue != ClusterBlockLevel.REFRESH || version.onOrAfter(TransportVersions.NEW_REFRESH_CLUSTER_BLOCK)
```

Comment:
This class should not have a reference to a specific enum value and transport version - this should be handled by the class doing the serialization

Reply:
I think there is no other way to do this check. Notice that the class
using this enum does do the right thing. I think we can either have the assertion or remove it. I would prefer to have it, similar to how we have the xcontenttype assertion.
We could have a separate task to introduce the infra to declare a min transport version on an enum to ensure this is verified in the infra.

Reply:
I added this assertion to ensure that we'll be informed in case I miss a place where the `ClusterBlockLevel` is serialized. I agree it's not great, but it helps me be more confident in the change as some part of this PR will be backported to 8.x (at least the new cluster block and the serialization changes).

I would prefer to keep this assertion until the backport is done and CI run for a couple of weeks before removing it, if that's possible.

Reply:
I think I'm with @thecoop here - `writeEnum` is only appropriate for the (very common) case that enum values map exactly to their ordinals in the wire protocol. In all other cases the caller should define its own mapping between enum values and wire representations, calling `writeVInt()` and `readVInt()` itself to deal with older versions that relied on `writeEnum()` and `readEnum()`.